packages/react-scheduler/src/ReactScheduler.js

```diff
@@ -300,11 +296,40 @@ if (!ExecutionEnvironment.canUseDOM) {
       isAnimationFrameScheduled = true;
       requestAnimationFrameForReact(animationTick);
     }
-    return newCallbackId;
+    return scheduledCallbackConfig;
   };
 
-  cancelScheduledWork = function(callbackId: number) {
-    delete registeredCallbackIds[callbackId];
+  cancelScheduledWork = function(
+    callbackConfig: CallbackIdType /* CallbackConfigType */,
+  ) {
+    /**
+     * There are four possible cases:
+     * - Head -> nodeToRemove -> Tail
+     *   In this case we point the Head.next to the Tail and the Tail.prev to
+     *   the Head.
+     * - Head -> middle -> Tail/nodeToRemove
+     *   In this case we point the middle.next to null and put middle as the new
+     *   Tail.
+     * - Head/nodeToRemove -> middle -> Tail
+     *   In this case we point the middle.prev at null and move the Head to
+     *   middle.
+     * - Head/nodeToRemove/Tail -> null
+     *   In this case we set Head and Tail to null.
+     */
+    const previousCallbackConfig = callbackConfig.previousCallbackConfig;
+    const nextCallbackConfig = callbackConfig.nextCallbackConfig;
+    if (previousCallbackConfig) {
```

Comment:
Same here, compare against `null`

Reply:
I find this much easier to follow using nested if statements. The reason I prefer that style is it makes it easier to avoid redundant type checks.

```js
if (callback === head) {
  if (callback === tail) {
    // List is now empty
    head = tail = null;
  } else {
    // Remove from start of list
    head = callback.next;
    // (Flow will complain that `head` could be null)
    head.previous = null;
  }
} else if (callback === tail) {
  // Remove from end of list
  tail = callback.previous;
  // (Flow will complain that `tail` could be null)
  tail.next = null;
} else {
  // Remove from middle of list
  const previous = callback.previous;
  const next = callback.next;
  // (Flow will complain that `previous` and `next` could be null)  
  previous.next = next;
  next.previous = previous;
}
```

But it seems like maybe the redundant checks are necessary to satisfy Flow? That's a shame.

Reply:
I can try to rewrite it this way, think we can still make Flow happy.
Might be ok with concise > clear in this particular module, but for now let me try the more clear way.